internal/controllers/machine/machine_controller_noderef.go

```diff
@@ -131,6 +141,57 @@ func (r *Reconciler) reconcileNode(ctx context.Context, cluster *clusterv1.Clust
 	return ctrl.Result{}, nil
 }
 
+// unstructuredNode returns a raw unstructured from a Node.
+func unstructuredNode(node *corev1.Node) *unstructured.Unstructured {
+	obj := &unstructured.Unstructured{}
+	obj.SetAPIVersion(node.String())
+	obj.SetKind(node.Kind)
+	obj.SetName(node.Name)
+	obj.SetUID(node.UID)
+	obj.SetLabels(node.Labels)
+	return obj
+}
+
+func reconcileManagedLabels(machine *clusterv1.Machine, node *corev1.Node) bool {
+	// Reconcile managed Labels.
+	var labelsHaveChanged bool
```

Comment:
I think we cannot calculate `labelsHaveChanged` without reimplementing SSA or using SSA dry-run.

(Assuming (as it's specified in the proposal) we only want to be authoritative for the labels which are currently set via the Machine or want to drop the ones that were previously set by us. Note: This excludes labels set by other users/controllers/kubeadm/...)

Essentially the current code calculates a diff and `labelsHaveChanged` is `true` either if there are additional labels on the Node or labels from the Machine are missing on the Node.

The problem is that it's a totally valid case that we have additional labels set by someone else on the Node.

Reply:
+1, if labels set via Machine are authoritative it totally makes sense to always add them to the node and let SSA handle everything else. The current implementation will be simplified by this approach.